File: exec/jumpexec.go
package exec

import (
	"bufio"
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"sync"

	"golang.org/x/sync/errgroup"
)

// JumpSpec errors.
var (
	ErrNoJumpCommand = errors.New("jump command not set")
	ErrNoJumpHosts   = errors.New("no jump hosts available")
)

// DefaultJumpCommand is the command used to access jump hosts.
const DefaultJumpCommand = "ssh -A -oBatchMode=yes -oConnectTimeout=10 {.J}"

// JumpSpec is the specification of a batch execution using jump hosts.
// Connection to jump hosts use ssh in batch mode.
//
// Each connection to a jump hosts spawns an ssh-agent to parallelize
// authentication. The JumpHostsKeyFile is automatically added to the
// ssh-agent via ssh-add if set.
type JumpSpec struct {
	Spec
	JumpHostsKeyFile string
	JumpCommand      string
	JumpHosts        []string
}

// Validate checks the spec. Returns error if required settings are not set.
func (s *JumpSpec) Validate() error {
	ss := &s.Spec
	if err := ss.Validate(); err != nil {
		return err
	}
	switch {
	case s.JumpCommand == "":
		return ErrNoJumpCommand
	case len(s.JumpHosts) == 0:
		return ErrNoJumpHosts
	}
	return nil
}

// JumpExec executes job s via jump hosts.
func JumpExec(ctx context.Context, w io.Writer, s *JumpSpec, targets <-chan string) error {
	if err := s.Validate(); err != nil {
		return err
	}

	parallel := s.Parallel / len(s.JumpHosts)
	if parallel == 0 {
		parallel = 1 // TODO: warn about too many jump hosts? cut some?
	}

	var mu sync.Mutex
	g, ctx := errgroup.WithContext(ctx)

	for _, host := range s.JumpHosts {
		// ssh-agent is a bottleneck; spawn one per jump host
		agentcmd, authsock, err := sshAgent(ctx, host)
		if err != nil {
			return err
		}

		if s.JumpHostsKeyFile != "" {
			err = sshAddKey(ctx, host, authsock, s.JumpHostsKeyFile)
			if err != nil {
				return fmt.Errorf("ssh-add failed: %v", err)
			}
		}

		// start ssh to jump host
		pr, pw := io.Pipe()
		stdout := &synchronizedWriter{Writer: w, Mutex: &mu}
		stderr := newLogWriter(ctx, host, true)

		cmd := strings.Replace(s.JumpCommand, "{.J}", host, -1)
		args := strings.Split(cmd, " ")
		cmd, args = args[0], args[1:]

		ushargs := []string{
			"--",
			"ush",
			"exec",
			"--timeout=" + s.Timeout.String(),
			"--parallel=" + strconv.Itoa(parallel),
			"--stdout_bytes=" + strconv.Itoa(s.StdoutBytes),
			"--stderr_bytes=" + strconv.Itoa(s.StderrBytes),
			"--",
			s.Command,
		}

		args = append(args, ushargs...)
		args = append(args, s.Args...)

		oscmd := exec.CommandContext(ctx, cmd, args...)
		oscmd.Stdin = pr
		oscmd.Stdout = stdout
		oscmd.Stderr = stderr
		oscmd.Env = []string{"SSH_AUTH_SOCK=" + authsock}

		err = oscmd.Start()
		if err != nil {
			return err
		}

		g.Go(func() error {
			defer pw.Close()
			for {
				select {
				case <-ctx.Done():
					return ctx.Err()
				case t, more := <-targets:
					if !more {
						return nil
					}
					_, err := io.WriteString(pw, t+"\n")
					if err != nil {
						return err
					}
				}
			}
		})

		g.Go(func() error {
			defer agentcmd.Process.Kill()
			defer stdout.Flush()
			return oscmd.Wait()
		})
	}

	return g.Wait()
}

type synchronizedWriter struct {
	io.Writer
	*sync.Mutex
	b bytes.Buffer
}

func (s *synchronizedWriter) Write(p []byte) (int, error) {
	idx := bytes.Index(p, []byte("\n"))
	switch idx {
	case -1:
		return s.b.Write(p)
	case 0:
		return len(p), nil
	}

	b := &s.b
	b.Write(p[:idx])

	s.Mutex.Lock()
	_, err := io.Copy(s.Writer, b)
	s.Mutex.Unlock()
	if err != nil {
		return 0, err
	}

	b.Reset()
	_, err = b.Write(p[idx:])

	return len(p), err
}

func (s *synchronizedWriter) Flush() error {
	s.Mutex.Lock()
	_, err := io.Copy(s.Writer, &s.b)
	s.Mutex.Unlock()
	return err
}

var errAgentTerminated = errors.New("ssh-agent terminated")

// starts an ssh agent, reads the first line from stdout and discard the rest.
// returns the agent's SSH_AUTH_SOCK.
func sshAgent(ctx context.Context, jumphost string) (*exec.Cmd, string, error) {
	authsock := make(chan string, 1)

	cmd := exec.CommandContext(ctx, "ssh-agent", "-D", "-s")
	cmd.Stdout = &sshAgentStdout{authsock: authsock}
	cmd.Stderr = newLogWriter(ctx, jumphost, true)
	err := cmd.Start()
	if err != nil {
		return nil, "", err
	}

	procErr := make(chan error, 1)
	go func() {
		procErr <- cmd.Wait()
		close(procErr)
	}()

	select {
	case s := <-authsock:
		return cmd, s, nil
	case <-ctx.Done():
		cmd.Process.Kill()
		return nil, "", errAgentTerminated
	case err := <-procErr:
		if err == nil {
			err = errAgentTerminated
		}
		return nil, "", err
	}
}

type sshAgentStdout struct {
	authsock chan string
	done     bool
	b        bytes.Buffer
}

func (s *sshAgentStdout) Write(p []byte) (int, error) {
	if s.done {
		return len(p), nil
	}

	// agent output: SSH_AUTH_SOCK=value; export SSH_AUTH_SOCKET\n

	idx := bytes.Index(p, []byte("\n"))
	n, err := s.b.Write(p)
	if err != nil || idx == -1 {
		return n, err
	}

	line1 := bytes.SplitN(s.b.Bytes(), []byte("\n"), 2)[0]
	parts := bytes.SplitN(line1, []byte("="), 2)
	if len(parts) != 2 && !bytes.Equal(parts[0], []byte("SSH_AUTH_SOCK")) {
		return 0, fmt.Errorf("unexpected output from ssh-agent: %q", string(line1))
	}

	sock := string(bytes.SplitN(parts[1], []byte(";"), 2)[0])
	if _, err := os.Stat(sock); err != nil {
		return 0, fmt.Errorf("cannot stat SSH_AUTH_SOCK: %v", err)
	}

	s.authsock <- sock
	s.done = true
	return len(p), nil
}

func sshAddKey(ctx context.Context, jumphost, authsock, authkey string) error {
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	cmd := exec.CommandContext(ctx, "ssh-add", authkey)
	cmd.Stdout = ioutil.Discard
	cmd.Stderr = newLogWriter(ctx, jumphost, false)
	cmd.Env = []string{"SSH_AUTH_SOCK=" + authsock}
	return cmd.Run()
}

func newLogWriter(ctx context.Context, jumphost string, isErr bool) io.Writer {
	pr, pw := io.Pipe()
	go func() {
		scanner := bufio.NewScanner(pr)
		for scanner.Scan() {
			fmt.Fprintf(os.Stderr, "%s: %s", jumphost, scanner.Text())
		}
	}()
	go func() {
		<-ctx.Done()
		pr.Close()
	}()
	return pw
}

File: src/exec/jumpexec.rs
use std::io::{BufRead, BufReader, Write};
use std::process::{Child, Command, Stdio};
use std::sync::{mpsc, Arc, Mutex};
use std::time::Duration;

use super::{is_shutdown, Spec};

pub const DEFAULT_JUMP_COMMAND: &str = "ssh -A -oBatchMode=yes -oConnectTimeout=10 -- {jump}";

pub struct JumpSpec {
    pub spec: Spec,
    pub jump_hosts_key_file: String,
    pub jump_command: String,
    pub jump_hosts: Vec<String>,
}

impl JumpSpec {
    pub fn validate(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.spec.validate()?;
        if self.jump_command.is_empty() {
            return Err("jump command not set".into());
        }
        if self.jump_hosts.is_empty() {
            return Err("no jump hosts available".into());
        }
        // Validate jump host names: reject whitespace and option-like strings
        for host in &self.jump_hosts {
            if host.contains(char::is_whitespace) {
                return Err(format!("jump host name contains whitespace: {:?}", host).into());
            }
            if host.starts_with('-') {
                return Err(format!("jump host name starts with dash: {:?}", host).into());
            }
        }
        Ok(())
    }
}

/// Executes commands via jump hosts with ssh-agent per host.
pub fn jump_exec(
    w: Arc<Mutex<Box<dyn Write + Send>>>,
    spec: &JumpSpec,
    input: mpsc::Receiver<String>,
) -> Result<(), Box<dyn std::error::Error>> {
    spec.validate()?;

    let parallel = std::cmp::max(1, spec.spec.parallel / spec.jump_hosts.len());
    let input = Arc::new(Mutex::new(input));
    let mut handles = Vec::new();

    vlog!(
        "[verbose] jump_exec: {} host(s), {} parallel per host",
        spec.jump_hosts.len(),
        parallel
    );

    for host in &spec.jump_hosts {
        let (mut agent_cmd, authsock) = ssh_agent(host)?;
        vlog!(
            "[verbose] jump_exec: ssh-agent started for {}, sock={}",
            host,
            authsock
        );

        if !spec.jump_hosts_key_file.is_empty() {
            ssh_add_key(host, &authsock, &spec.jump_hosts_key_file)?;
            vlog!("[verbose] jump_exec: key loaded for {}", host);
        }

        // Build SSH command: split the template, substitute {jump} into individual args
        // This prevents argument injection via host names containing spaces or -o flags
        let template_parts: Vec<&str> = spec.jump_command.split_whitespace().collect();
        if template_parts.is_empty() {
            return Err("empty jump command".into());
        }

        let ssh_cmd = template_parts[0].to_string();
        let mut args: Vec<String> = template_parts[1..]
            .iter()
            .map(|s| s.replace("{jump}", host))
            .collect();

        // Append the remote ush command after the SSH args
        args.push("ush".to_string());
        args.push("exec".to_string());
        args.push(format!(
            "--timeout={}",
            crate::time::format_go_duration(spec.spec.timeout)
        ));
        args.push(format!("--parallel={}", parallel));
        args.push(format!("--stdout_bytes={}", spec.spec.stdout_bytes));
        args.push(format!("--stderr_bytes={}", spec.spec.stderr_bytes));
        if spec.spec.head {
            args.push("--head".to_string());
        }
        args.push("--".to_string());
        args.push(spec.spec.command.clone());
        args.extend(spec.spec.args.clone());

        vlog!(
            "[verbose] jump_exec: spawning {} {:?} for host {}",
            ssh_cmd,
            args,
            host
        );

        let mut child = Command::new(&ssh_cmd)
            .args(&args)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .env_clear()
            .env("SSH_AUTH_SOCK", &authsock)
            .spawn()?;

        let child_stdin = child.stdin.take().unwrap();
        let child_stdout = child.stdout.take().unwrap();
        let child_stderr = child.stderr.take().unwrap();

        let input = input.clone();
        let w = w.clone();
        let host_name = host.clone();

        // Feed targets to stdin
        let feed_handle = std::thread::spawn(move || {
            let mut stdin = child_stdin;
            loop {
                if is_shutdown() {
                    break;
                }
                let target = {
                    let rx = input.lock().unwrap();
                    rx.recv().ok()
                };
                match target {
                    Some(t) => {
                        if writeln!(stdin, "{}", t).is_err() {
                            break;
                        }
                    }
                    None => break,
                }
            }
            drop(stdin);
        });

        // Read stdout (JSON lines) and write synchronized
        let w2 = w.clone();
        let stdout_handle = std::thread::spawn(move || {
            let reader = BufReader::new(child_stdout);
            for line in reader.lines() {
                match line {
                    Ok(line) => {
                        if line.is_empty() {
                            continue;
                        }
                        let mut w = w2.lock().unwrap();
                        let _ = writeln!(&mut *w, "{}", line);
                    }
                    Err(_) => break,
                }
            }
        });

        // Read stderr and log
        let host_name2 = host_name.clone();
        let stderr_handle = std::thread::spawn(move || {
            let reader = BufReader::new(child_stderr);
            for line in reader.lines() {
                match line {
                    Ok(line) => {
                        eprintln!("{}: {}", host_name2, line);
                    }
                    Err(_) => break,
                }
            }
        });

        handles.push(std::thread::spawn(move || {
            let _ = feed_handle.join();
            let _ = stdout_handle.join();
            let _ = stderr_handle.join();
            let _ = child.wait();
            vlog!(
                "[verbose] jump_exec: SSH to {} finished, killing ssh-agent",
                host_name
            );
            let _ = agent_cmd.kill();
            let _ = agent_cmd.wait();
        }));
    }

    for handle in handles {
        let _ = handle.join();
    }

    Ok(())
}

fn ssh_agent(jumphost: &str) -> Result<(Child, String), Box<dyn std::error::Error>> {
    let mut cmd = Command::new("ssh-agent")
        .args(["-D", "-s"])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;

    let stdout = cmd.stdout.take().unwrap();
    let mut reader = BufReader::new(stdout);
    let mut first_line = String::new();
    reader.read_line(&mut first_line)?;

    // Parse SSH_AUTH_SOCK=value; export SSH_AUTH_SOCK;
    let parts: Vec<&str> = first_line.splitn(2, '=').collect();
    if parts.len() != 2 {
        let _ = cmd.kill();
        return Err(format!(
            "{}: unexpected output from ssh-agent: {:?}",
            jumphost, first_line
        )
        .into());
    }

    let sock = parts[1].split(';').next().unwrap_or("").to_string();
    if sock.is_empty() {
        let _ = cmd.kill();
        return Err(format!("{}: empty SSH_AUTH_SOCK from ssh-agent", jumphost).into());
    }

    // Retry socket existence check with short timeout
    let mut found = false;
    for attempt in 0..20 {
        if std::path::Path::new(&sock).exists() {
            found = true;
            break;
        }
        vlog!(
            "[verbose] ssh_agent: waiting for socket {} (attempt {})",
            sock,
            attempt + 1
        );
        std::thread::sleep(Duration::from_millis(100));
    }
    if !found {
        let _ = cmd.kill();
        return Err(format!(
            "{}: ssh-agent socket not found after 2s: {}",
            jumphost, sock
        )
        .into());
    }

    Ok((cmd, sock))
}

fn ssh_add_key(
    jumphost: &str,
    authsock: &str,
    keyfile: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let output = Command::new("ssh-add")
        .arg(keyfile)
        .env_clear()
        .env("SSH_AUTH_SOCK", authsock)
        .stdout(Stdio::null())
        .stderr(Stdio::piped())
        .output()?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("{}: ssh-add failed: {}", jumphost, stderr.trim()).into());
    }

    Ok(())
}
